### ISSUE TITLE
Building Trilinos with Kokkos enabled doesn't work on 32-bit machines

### ISSUE THREAD

**Issue by BarrySmith** [trilinos/Trilinos#82]
The PETSc ./configure --download-trilinos=1 --with-cxx-dialect=C++11 --download-boost=1 --download-cmake=1 

is failing on one of our machines with error message: 

In file included from /sandbox/sarich/petsc/arch-linux2-c-debug/externalpackages/git.trilinos/packages/kokkos/core/src/Kokkos_Atomic.hpp:208:0,
                 from /sandbox/sarich/petsc/arch-linux2-c-debug/externalpackages/git.trilinos/packages/kokkos/core/src/impl/Kokkos_AllocationTracker.cpp:50:
/sandbox/sarich/petsc/arch-linux2-c-debug/externalpackages/git.trilinos/packages/kokkos/core/src/impl/Kokkos_Atomic_Generic.hpp: In instantiation of ‘T Kokkos::Impl::atomic_fetch_oper(const Oper&, volatile T_, typename Kokkos::Impl::enable_if<((sizeof (T) != sizeof (int)) && (sizeof (T) == sizeof (long long unsigned int))), const T>::type) [with Oper = Kokkos::Impl::AndOper<long long unsigned int, const long long unsigned int>; T = long long unsigned int; typename Kokkos::Impl::enable_if<((sizeof (T) != sizeof (int)) && (sizeof (T) == sizeof (long long unsigned int))), const T>::type = const long long unsigned int]’:
/sandbox/sarich/petsc/arch-linux2-c-debug/externalpackages/git.trilinos/packages/kokkos/core/src/impl/Kokkos_Atomic_Generic.hpp:296:69:   required from ‘T Kokkos::atomic_fetch_and(volatile T_, T) [with T = long long unsigned int]’
/sandbox/sarich/petsc/arch-linux2-c-debug/externalpackages/git.trilinos/packages/kokkos/core/src/impl/Kokkos_AllocationTracker.cpp:211:64:   required from ‘bool Kokkos::Impl::{anonymous}::Bitset<NumBlocks>::reset(int) [with int NumBlocks = 15]’
/sandbox/sarich/petsc/arch-linux2-c-debug/externalpackages/git.trilinos/packages/kokkos/core/src/impl/Kokkos_AllocationTracker.cpp:296:30:   required from here
/sandbox/sarich/petsc/arch-linux2-c-debug/externalpackages/git.trilinos/packages/kokkos/core/src/impl/Kokkos_Atomic_Compare_Exchange_Strong.hpp:207:3: note: template<class T> T Kokkos::atomic_compare_exchange(volatile T_, T, typename Kokkos::Impl::enable_if<((sizeof (T) != 4) && (sizeof (T) != 8)), const T>::type&)
 T atomic_compare_exchange( volatile T \* const dest , const T compare ,
   ^
/sandbox/sarich/petsc/arch-linux2-c-debug/externalpackages/git.trilinos/packages/kokkos/core/src/impl/Kokkos_Atomic_Compare_Exchange_Strong.hpp:207:3: note:   template argument deduction/substitution failed:
/sandbox/sarich/petsc/arch-linux2-c-debug/externalpackages/git.trilinos/packages/kokkos/core/src/impl/Kokkos_Atomic_Compare_Exchange_Strong.hpp: In substitution of ‘template<class T> T Kokkos::atomic_compare_exchange(volatile T_, T, typename Kokkos::Impl::enable_if<((sizeof (T) != 4) && (sizeof (T) != 8)), const T>::type&) [with T = long long unsigned int]’:
/sandbox/sarich/petsc/arch-linux2-c-debug/externalpackages/git.trilinos/packages/kokkos/core/src/impl/Kokkos_Atomic_Generic.hpp:144:103:   required from ‘T Kokkos::Impl::atomic_fetch_oper(const Oper&, volatile T_, typename Kokkos::Impl::enable_if<((sizeof (T) != sizeof (int)) && (sizeof (T) == sizeof (long long unsigned int))), const T>::type) [with Oper = Kokkos::Impl::AndOper<long long unsigned int, const long long unsigned int>; T = long long unsigned int; typename Kokkos::Impl::enable_if<((sizeof (T) != sizeof (int)) && (sizeof (T) == sizeof (long long unsigned int))), const T>::type = const long long unsigned int]’
/sandbox/sarich/petsc/arch-linux2-c-debug/externalpackages/git.trilinos/packages/kokkos/core/src/impl/Kokkos_Atomic_Generic.hpp:296:69:   required from ‘T Kokkos::atomic_fetch_and(volatile T_, T) [with T = long long unsigned int]’
/sandbox/sarich/petsc/arch-linux2-c-debug/externalpackages/git.trilinos/packages/kokkos/core/src/impl/Kokkos_AllocationTracker.cpp:211:64:   required from ‘bool Kokkos::Impl::{anonymous}::Bitset<NumBlocks>::reset(int) [with int NumBlocks = 15]’
/sandbox/sarich/petsc/arch-linux2-c-debug/externalpackages/git.trilinos/packages/kokkos/core/src/impl/Kokkos_AllocationTracker.cpp:296:30:   required from here
/sandbox/sarich/petsc/arch-linux2-c-debug/externalpackages/git.trilinos/packages/kokkos/core/src/impl/Kokkos_Atomic_Compare_Exchange_Strong.hpp:207:3: error: invalid use of incomplete type ‘struct Kokkos::Impl::enable_if<false, const long long unsigned int>’
In file included from /sandbox/sarich/petsc/arch-linux2-c-debug/externalpackages/git.trilinos/packages/kokkos/core/src/Kokkos_MemoryTraits.hpp:47:0,
                 from /sandbox/sarich/petsc/arch-linux2-c-debug/externalpackages/git.trilinos/packages/kokkos/core/src/Kokkos_HostSpace.hpp:53,
                 from /sandbox/sarich/petsc/arch-linux2-c-debug/externalpackages/git.trilinos/packages/kokkos/core/src/Kokkos_Atomic.hpp:71,
                 from /sandbox/sarich/petsc/arch-linux2-c-debug/externalpackages/git.trilinos/packages/kokkos/core/src/impl/Kokkos_AllocationTracker.cpp:50:
/sandbox/sarich/petsc/arch-linux2-c-debug/externalpackages/git.trilinos/packages/kokkos/core/src/impl/Kokkos_Traits.hpp:201:8: error: declaration of ‘struct Kokkos::Impl::enable_if<false, const long long unsigned int>’
 struct enable_if ;
        ^

Full output from configure/build is attached

[configure.txt](https://github.com/trilinos/Trilinos/files/95073/configure.txt)

Barry
@sarich
@balay@mcs.anl.gov

**Comment by hcedwar** [comment]
Compiler appears to be GCC 4.6.3, must have at least 4.7.2.

**Comment by BarrySmith** [comment]
> On Jan 18, 2016, at 4:29 PM, hcedwar <notifications@github.com> wrote:
> 
> Compiler appears to be GCC 4.6.3, must have at least 4.7.2.

   Then you are damn well obligated to have your Cmake code check the compiler version and verify that it is good enough! You cannot have these errors pop up at make time when they should be checked at Cmake time.

**Comment by balay** [comment]
from configure.log

Executing: mpicxx --version
stdout:
g++ (GCC) 4.8.0

Will check with 4.9.0

**Comment by balay** [comment]
Also Jason's build is on a 32bit machine. Another variable to check..

from configure.log

                Defined "SIZEOF_VOID_P" to "4"

**Comment by balay** [comment]
I have a successful build with gcc-4.8.0 and 4.9.0 on linux64  [also with 5.3.1on a different linux64 box] - so I suspect its a code issue on 32bit builds..

I'm attempting a build with gcc-4.8.4 on a different linux32 box. Will see how that goes..

**Comment by BarrySmith** [comment]
Newer versions of CMake (2.8.11+) supports giving you the compiler version for C and C++.  See:

   https://cmake.org/cmake/help/v2.8.11/cmake.html#variable:CMAKE_LANG_COMPILER_VERSION

In fact, TriBITS already was printing this, for example on my machine:

...
-- The CXX compiler identification is GNU 4.8.3                                                                                                                 
-- Check for working CXX compiler: /projects/vera/gcc-4.8.3/toolset/mpich-3.1.3/bin/mpicxx                                                                      
-- Check for working CXX compiler: /projects/vera/gcc-4.8.3/toolset/mpich-3.1.3/bin/mpicxx -- works                                                             
-- Detecting CXX compiler ABI info                                                                                                                              
-- Detecting CXX compiler ABI info - done                                                                                                                       
-- CMAKE_CXX_COMPILER_ID='GNU'                                                                                                                                  
-- CMAKE_CXX_COMPILER_VERSION='4.8.3'
...


Just put in something like the following in your kokkos/CMakeLists.txt file:

  IF (CMAKE_C_COMPILER_ID STREQUAL "GNU" AND
    CMAKE_CXX_COMPILER_VERSION VERSION_LESS "4.7.2"
    )
    MESSAGE(FATAL_ERROR "Error, Kokkos requires GCC 4.7.2 or greater, but was only"
      " provided GCC ${ CMAKE_CXX_COMPILER_VERSION}")
  ENDIF()

Cheers,

-Ross


> -----Original Message-----
> From: Barry Smith [mailto:bsmith@mcs.anl.gov]
> Sent: Monday, January 18, 2016 5:33 PM
> To: trilinos/Trilinos; Willenbring, James M; Heroux, Michael A; Bartlett,
> Roscoe A.
> Cc: trilinos/Trilinos; Jason Sarich; Satish Balay
> Subject: Re: [Trilinos] Failed build of Trilinos off of master due to Kokkos
> issues (#82)
> 
>

**Comment by crtrott** [comment]
For me the error looks like a real bug in Kokkos with respect to 32bit machines. Unfortunately 32bit is not anywhere in either our Kokkos standalone testing suite (which otherwise is pretty comprehensive including more than 10 compilers) nor in the Trilinos testing environment. That means we don't catch those errors. The interesting question for Trilinos is whether we have/keep a promise to work on 32 bit.

**Comment by crtrott** [comment]
The CMake logic thing for checking compiler is interesting I didn't know that that was added. I believe we require 2.8.11 anyway, so we can add that part - but as I said I don't believe its the problem in this case.

**Comment by maherou** [comment]
There is certainly a lot less interest in 32-bit (global ordinal) types from our main users.  32-bit local ordinals are still very important and have significant performance impact.  We have decided not to permit int global ordinals for our explicit template instantiation builds, but there are still instances where you can argue for 32.  No matter what, we should give clear error messages if we are not going to support it.

**Comment by bartlettroscoe** [comment]
TriBITS requires CMake 2.8.11+ so it is safe to put in the exact logic I gave (just make sure you test it first).

**Comment by crtrott** [comment]
This is not types, mike this is the build type (do you build for a 32 bit machine or a 64 bit machine) which affects things like size of pointer, or size of long.

**Comment by maherou** [comment]
Sorry, I missed that.  Wow, I don't see us needing to support 32-bit pointers.

**Comment by crtrott** [comment]
The thing is we have to do either or: either we support 32 bit systems, and then we have to somehow get tests for that set up, or we should fatal error when configuring for 32 bit. Myself I would be more happy with the latter, since that means one less headache for me. But that is probably a discussion for the leadership phone call on Thursday.

**Comment by hcedwar** [comment]
Need a configure test with 'static_assert( sizeof(void*) == 8 )'

**Comment by maherou** [comment]
I have added it as an item to the leaders agenda for Thursday.

**Comment by BarrySmith** [comment]
I am fine with not having support in xSDK for 32 bit builds, just let me know what you decide and we can document it in the xSDK google doc. 

   I also apologize for using damn in my bug report. I should have just properly stated I recommend adding a cmake/configure test for this requirement instead of trying to be colorful but instead just being stupid.

   Barry

**Comment by mhoemmen** [comment]
@maherou wrote: "We have decided not to permit int global ordinals for our explicit template instantiation builds..."  That is not accurate.  Making it _possible_ _to_ _disable_ `GlobalOrdinal = int` AT ALL (regardless of ETI) has taken quite a bit of work by several Trilinos developers in many packages since Sep 2015.  This started in Tpetra but ended up taking a lot of effort in Xpetra, MueLu, Amesos2, Zoltan2, and other packages.  Internal customers are just now starting to experiment with this capability, and it's still a bit iffy.

**Comment by maherou** [comment]
OK.  Let's get to an accurate statement.  It is my understanding that, once the issues related to turning off GlobalOrdinal = int (which includes issues in Sierra as well as Trilinos, and probably other apps too), we will not include GlobalOrdinal = int in the default list of ETI types.  Is this a true statement?  If not, please rephrase to make it correct.  I want to make sure I understand the plans for this.

**Comment by mhoemmen** [comment]
@maherou Hi Mike!  Hope I'm not leading this thread on a tangent.  The current plan is to make sure that Sierra and possibly other apps can turn off `GlobalOrdinal=int` and still get the functionality they need.  Once this is fully achieved (it's pretty much there now), we can talk about turning it off _by_ _default_.  However, doing that while still getting full code and test coverage would take more work.

In particular, support for Epetra in Xpetra (and its dependencies, like MueLu and Zoltan2) and Amesos2 depends on `GO=int`.  If we turn that off by default, we won't get test coverage for Epetra interfaces in MueLu and Amesos2 by default.  Changing this would require refactoring Xpetra and Amesos2 to replace the Epetra wrappers with stub (or GO conversion) layers when `GO=int` is OFF.  This would take a lot of work and be very disruptive to both the Amesos2 and MueLu teams, at a critical time for them.  I estimate 3-4 FTE effort for 6 weeks, plus disruption of other developers due to configuration issues.

Thus, we haven't committed to changing the default setting.  I favor the _capability_ of turning it off and have invested significant personal and social effort in making this work.  We are working with Mike Glass on this actively (I've been answering a lot of e-mails from him over the past few days).  However, changing the _default_ setting isn't for me to decide, given the effort it would take.

To summarize: It's possible now to turn off `GO=int` in many packages, including all those that Sierra currently uses.  However, making this the _default_ setting would take a lot more work.

**Comment by crtrott** [comment]
Guys this problem has nothing to do with the Global or Local ordinal size. This is a 32 bit build type vs 64 build type (as in 32 bit machine vs 64 bit machine)

**Comment by BarrySmith** [comment]
We've changed the  PETSc build for Trilinos to error out with a very useful message if someone attempts a PETSc build with Trilinos and 32 bit pointers. Hence you will not get error reports like this in the future.

   Barry

**Comment by mhoemmen** [comment]
I changed the bug title to reflect the actual issue.
